pkg/kubelet/logs/container_log_manager.go

```diff
@@ -213,63 +223,110 @@ func (c *containerLogManager) Clean(ctx context.Context, containerID string) err
 	if err != nil {
 		return fmt.Errorf("failed to list containers: %v", err)
 	}
-	// NOTE(random-liu): Figure out whether we need to rotate container logs in parallel.
 	for _, container := range containers {
 		// Only rotate logs for running containers. Non-running containers won't
 		// generate new output, it doesn't make sense to keep an empty latest log.
 		if container.GetState() != runtimeapi.ContainerState_CONTAINER_RUNNING {
 			continue
 		}
-		id := container.GetId()
-		// Note that we should not block log rotate for an error of a single container.
-		resp, err := c.runtimeService.ContainerStatus(ctx, id, false)
-		if err != nil {
-			klog.ErrorS(err, "Failed to get container status", "containerID", id)
-			continue
-		}
-		if resp.GetStatus() == nil {
-			klog.ErrorS(err, "Container status is nil", "containerID", id)
-			continue
+		// Doing this to avoid additional overhead with logging of label like arguments that can prove costly
+		if v := klog.V(4); v.Enabled() {
+			klog.V(4).InfoS("Adding new entry to the queue for processing", "id", container.GetId(), "name", container.Metadata.GetName(), "labels", container.GetLabels())
 		}
-		path := resp.GetStatus().GetLogPath()
-		info, err := c.osInterface.Stat(path)
-		if err != nil {
-			if !os.IsNotExist(err) {
-				klog.ErrorS(err, "Failed to stat container log", "path", path)
-				continue
-			}
-			// In rotateLatestLog, there are several cases that we may
-			// lose original container log after ReopenContainerLog fails.
-			// We try to recover it by reopening container log.
-			if err := c.runtimeService.ReopenContainerLog(ctx, id); err != nil {
-				klog.ErrorS(err, "Container log doesn't exist, reopen container log failed", "containerID", id, "path", path)
-				continue
-			}
-			// The container log should be recovered.
-			info, err = c.osInterface.Stat(path)
-			if err != nil {
-				klog.ErrorS(err, "Failed to stat container log after reopen", "path", path)
-				continue
-			}
+		c.queue.Add(container.GetId())
+	}
+
+	// There are two reasons why we want to wait for this queue to be empty before we return from this function
+	// and let the next iteration of the rotation start.
+	//
+	// 1. When the default values are used for the max number of container log rotation workers and the monitoring
+	//    duration, we want to make sure the original behavior of the current log rotation is retained. Which was
+	//    processing the contaienrs in a loop and then starting the next iteration. Waiting for the queue to be
+	//    emopty lets us get the same behavior.
+	//
+	// 2. When there is more than one container log rotation workfer is running, If we do not wait for the queue to be empty
+	//    It can lead to race condition between the log rotation workflow when they get picked up by two different goroutines
+	//    which can lead to inconsistent behavior on the rotation. Waiting for the queue to be empty helps us mititage that
```

Comment:
The inconsistency I was thinking about was, If we do not wait for the queue to drain and return out of the loop, the next iteration of the workflow for rotate can kick in. And when that happens, if the item that gets picked up by one of the queue is still being processed for the previous item in a different queue, now we have a case where the same container log is being rotated by two different worker routines because we did not wait for the first queue to be drained entirely. The chances of this race happening might be less but very much possible. 

That being said, I see what you mean. The workquee's `.Add(item)` API has a `q.dirty.has(item)` check. Do you think this is enough of validation to avoid race condition between two workers picking up the same Container and start rotating the same set of logs ? If so, I can get rid of the wait check for queue drain and run a few tests and push the changes to the PR>

Reply:
LI1. WaitForver -> rotateLogs -> Add to queue -> Returns back (Worker routines will pick up and start processing each)
LI2. WaitForever kicks in again -> rotateLogs -> Add to queue -> `Add(item)` check if the entry is in dirty queue. If not add it -> Check the processing queue -> If so return. So this will not have a duplicate entry being processed in any other worker.
LI3. WaitForever kicks in again -> rotateLogs -> Add to queue -> `Add(item)` check if the entry is in dirty queue -> Finds it and returns right away. Avoid being processed again

Once the item from LI1 is complete, the queue will check for the dirty, find the entry from LI2 and add it back to the queue. 

Thanks for the suggestion @liggitt  I think we can avoid the wait loop for the queue drain based on the above sequence. Did I miss anything else ?

Reply:
Exactly, the workqueue holds a requeued item in the dirty queue until the current handler is done processing it, before it moves it to the main queue for any arbitrary worker to pick up

Reply:
Doing this would mean I might have to fix a few things on the tests and add a loop to wait for the queue to drain. Otherwise, the checks might start failing as the call to `rotateLogs` might return before the rotation of the logs is actually completed.

Reply:
I think the change to do this introduced a race and that is seemingly breaking and e2e test on gce.. Trying to confirm.